cdc_iterator.go

```diff
@@ -0,0 +1,243 @@
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/conduitio-labs/conduit-connector-mysql/common"
+	"github.com/conduitio/conduit-commons/csync"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/go-mysql-org/go-mysql/canal"
+	"github.com/go-mysql-org/go-mysql/mysql"
+	"github.com/go-mysql-org/go-mysql/schema"
+	mysqldriver "github.com/go-sql-driver/mysql"
+)
+
+type cdcIterator struct {
+	canal *canal.Canal
+	acks  *csync.WaitGroup
+	data  chan *canal.RowsEvent
+
+	canalRunErrC chan error
+	canalDoneC   chan struct{}
+
+	config cdcIteratorConfig
+}
+
+type cdcIteratorConfig struct {
+	tables      []string
+	mysqlConfig *mysqldriver.Config
+	position    sdk.Position
+	TableKeys   common.TableKeys
+}
+
+func newCdcIterator(ctx context.Context, config cdcIteratorConfig) (common.Iterator, error) {
+	c, err := common.NewCanal(config.mysqlConfig, config.tables)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create canal: %w", err)
+	}
+
+	sdk.Logger(ctx).Info().Msg("created canal")
+
+	iterator := &cdcIterator{
+		canal:        c,
+		acks:         &csync.WaitGroup{},
+		data:         make(chan *canal.RowsEvent),
+		canalRunErrC: make(chan error),
+		canalDoneC:   make(chan struct{}),
+		config:       config,
+	}
+
+	startPosition, err := iterator.getStartPosition(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get start position: %w", err)
+	}
+
+	go func() {
+		handler := &cdcEventHandler{
+			canalDoneC: iterator.canalDoneC,
+			data:       iterator.data,
+		}
+		iterator.canal.SetEventHandler(handler)
+
+		iterator.canalRunErrC <- iterator.canal.RunFrom(startPosition)
+	}()
+
+	return iterator, nil
+}
+
+func (c *cdcIterator) getStartPosition(config cdcIteratorConfig) (mysql.Position, error) {
+	if config.position != nil {
+		pos, err := common.ParseSDKPosition(config.position)
+		if err != nil {
+			return mysql.Position{}, fmt.Errorf("failed to parse position: %w", err)
+		}
+		if pos.Kind == common.PositionTypeSnapshot {
+			return mysql.Position{}, fmt.Errorf("invalid position type: %s", pos.Kind)
+		}
+
+		return pos.CdcPosition.Position, nil
+	}
+
+	masterPos, err := c.canal.GetMasterPos()
+	if err != nil {
+		return mysql.Position{}, fmt.Errorf("failed to get master position: %w", err)
+	}
+
+	return masterPos, nil
+}
+
+func (c *cdcIterator) Ack(context.Context, sdk.Position) error {
+	c.acks.Done()
+	return nil
+}
+
+func (c *cdcIterator) Next(ctx context.Context) (rec sdk.Record, err error) {
+	select {
+	case <-ctx.Done():
+		//nolint:wrapcheck // no need to wrap canceled error
+		return rec, ctx.Err()
+	case <-c.canalDoneC:
+		return rec, fmt.Errorf("canal is closed")
+	case data := <-c.data:
+		rec, err := c.buildRecord(data)
+		if err != nil {
+			return rec, fmt.Errorf("failed to build record: %w", err)
+		}
+
+		c.acks.Add(1)
+		return rec, nil
+	}
+}
+
+func (c *cdcIterator) Teardown(ctx context.Context) error {
+	close(c.canalDoneC)
+
+	c.canal.Close()
+	if err := <-c.canalRunErrC; err != nil {
```

Comment:
Would be nice to also listen to `ctx.Done` in parallel, otherwise we risk potentially getting stuck forever.

Reply:
You mean simultaneously right? in a `select` statement?